contracts/src/Cids.sol

```diff
@@ -7,34 +7,49 @@ library Cids {
         bytes data;
     }
 
+    // Returns the last 32 bytes of a CID payload as a bytes32.
     function digestFromCid(Cid memory cid) public pure returns (bytes32) {
         require(cid.data.length >= 32, "Cid data is too short");
-        // bytes memory dataSlice = new bytes(32);
-        // Assembly version of
-        // for (uint i = 0; i < 32; i++) {
-        //     dataSlice[i] = cid.data[cid.data.length - 32 + i];
-        // }
-        // return bytes32(dataSlice);   
-        assembly {
-            // Allocate memory for the data slice.
-            // Set dataSclice to the next 32-byte aligned location after the free memory pointer (at 0x40).
-            let dataSlice := add(mload(0x40), 0x20)
-            // Update free memory pointer to next slot after dataSlice.
-            mstore(0x40, add(dataSlice, 0x20))
+        bytes memory dataSlice = new bytes(32);
+        for (uint i = 0; i < 32; i++) {
+            dataSlice[i] = cid.data[cid.data.length - 32 + i];
+        }
+        return bytes32(dataSlice);   
+        // The above can almost certainly be optimised in assembly.
+        // However, the below assembly behaves differently and is beyond my ability to resolve.
```

Comment:
@ZenGround0 I tried assembly after your suggestion in the previous PR, but eventually tracked proof failures down to this. I would rather leave it to an expert. I can delete the incorrect assembly if you wish, though it might provide a head start to someone coming in.

Reply:
Its way better to keep it commented out, thank you